kimchi/src/folding/error.rs

```diff
@@ -0,0 +1,141 @@
+            EvalLeaf::bin_op(a, b, |a, b| *a + b, |a, b| *a += b)
+        }
+        FoldingExp::Sub(e1, e2) => {
+            let a = eval_exp_error(e1, env, side);
+            let b = eval_exp_error(e2, env, side);
+            EvalLeaf::bin_op(a, b, |a, b| *a - b, |a, b| *a -= b)
+        }
+        FoldingExp::Mul(e1, e2) => match (degree, e1.degree()) {
+            (Degree::Two, Degree::One) => {
+                let a = eval_exp_error(e1, env, side);
+                let b = eval_exp_error(e2, env, !side);
+                let first = EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b);
+                let a = eval_exp_error(e1, env, !side);
+                let b = eval_exp_error(e2, env, side);
+                let second = EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b);
+                EvalLeaf::bin_op(first, second, |a, b| *a + b, |a, b| *a += b)
+            }
+            _ => {
+                let a = eval_exp_error(e1, env, side);
+                let b = eval_exp_error(e2, env, side);
+                EvalLeaf::bin_op(a, b, |a, b| *a * b, |a, b| *a *= b)
+            }
+        },
+        FoldingExp::UnnormalizedLagrangeBasis(i) => EvalLeaf::Result(env.lagrange_basis(i)),
+        //TODO: use cache
+        FoldingExp::Cache(_id, exp) => eval_exp_error(exp, env, side),
+    }
+}
+
+pub(crate) fn compute_error<F: Field, E: FoldingEnv<F>>(
+    exp: &IntegratedFoldingExpr<F>,
+    env: &E,
+    u: (F, F),
+) -> Vec<F> {
+    let add = |a, b| EvalLeaf::bin_op(a, b, |a, b| *a + b, |a, b| *a += b);
+    let sub = |a, b| EvalLeaf::bin_op(a, b, |a, b| *a - b, |a, b| *a -= b);
+    let scale = |t, s| EvalLeaf::bin_op(t, EvalLeaf::Const(s), |a, b| *a * b, |a, b| *a *= b);
+
+    let t_0 = EvalLeaf::Result(env.zero_vec());
+    let t_0 = exp.degree_0.iter().fold(t_0, |t_0, (exp, sign)| {
+        //could be true or false, doesn't matter for constant terms
+        let e = eval_exp_error(exp, env, Side::Left);
+        if *sign {
+            add(t_0, e)
+        } else {
+            sub(t_0, e)
+        }
+    });
+    let t_0 = scale(t_0, (u.0 * u.1).double());
+
+    let t_1l = EvalLeaf::Result(env.zero_vec());
+    let t_1r = EvalLeaf::Result(env.zero_vec());
+    let (t_1l, t_1r) = exp
+        .degree_0
+        .iter()
+        .fold((t_1l, t_1r), |(t_1l, t_1r), (exp, sign)| {
+            let el = eval_exp_error(exp, env, Side::Left);
+            let er = eval_exp_error(exp, env, Side::Right);
+            if *sign {
+                (add(t_1l, el), add(t_1r, er))
+            } else {
+                (sub(t_1l, el), sub(t_1r, er))
+            }
+        });
+    let t_1l = scale(t_1l, u.1);
+    let t_1r = scale(t_1r, u.0);
+    let t_1 = EvalLeaf::bin_op(t_1l, t_1r, |a, b| *a + b, |a, b| *a += b);
+
+    let t_2 = EvalLeaf::Result(env.zero_vec());
+    let t_2 = exp.degree_0.iter().fold(t_2, |t_2, (exp, sign)| {
+        //true or false matter in some way, but not at the top level call
```

Comment:
Here again, Left Right

Reply:
In any case, this comment is a bit hand-wavy and does not contain much information an outsider can understand.